src/local.rs

```diff
@@ -1,45 +1,184 @@
-        let idx = unsafe {
-            // safety: hopefully, the provided `Locality` upholds its end of the
-            // contract...
-            L::current().into_usize()
-        };
-        unimplemented!()
-        // let item = if let Some(i) = unsafe { (*self.items[idx].get()).as_ref() } {
-        //     i
-        // } else {
-        //     let ptr = self.items[idx].get();
-        //     unsafe {
-        //         (*ptr) = Some((self.init)());
-        //         (*ptr).as_ref().expect("we just set the pointed value!")
-        //     }
-        // };
-        // f(item)
+#[cfg(not(feature = "std"))]
+#[doc(hidden)]
+#[macro_export]
+macro_rules! __locality_inner {
+    (@key $(#[$attr:meta])* $vis:vis $name:ident, $t:ty, $init:expr) => {
+        {
+            #[inline]
+            fn __init() -> $t { $init }
+
+            unsafe fn __getit() -> Option<&'static $t> {
+                static __KEY: $crate::local::__Store<$t> =
+                    $crate::local::__Store::new();
+
+                __KEY.get(__init)
+            }
+
+            unsafe {
+                $crate::local::LocalKey::new(__getit)
+            }
+        }
+    };
+
+    ($(#[$attr:meta])* $vis:vis static $name:ident, $t:ty, $init:expr) => {
+        $(#[$attr])* $vis const $name: $crate::local::LocalKey<$t> =
+            $crate::__locality_inner!(@key $(#[$attr])* $vis $name, $t, $init);
+    };
+}
+
+#[cfg(not(feature = "std"))]
+mod no_std {
+    use crate::stdlib::cell::UnsafeCell;
+    use crate::stdlib::{
+        fmt, mem, ptr,
+        sync::atomic::{AtomicPtr, AtomicU8, AtomicUsize, Ordering},
+    };
+    use crate::{Id, Locality};
+
+    pub(crate) static MAX_LOCALITIES: AtomicUsize = AtomicUsize::new(0);
+
+    static DEFAULT_LOCALITY: AtomicPtr<()> = AtomicPtr::new(ptr::null_mut());
+
+    pub struct SetDefaultError {
+        msg: &'static str,
+    }
+
+    impl fmt::Debug for SetDefaultError {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            write!(f, "SetDefaultError: {}", self.msg)
+        }
+    }
+
+    pub fn set_locality<L: Locality>() -> Result<(), SetDefaultError> {
+        let max = L::MAX_LOCALITIES;
+        MAX_LOCALITIES
+            .compare_exchange(0, max, Ordering::AcqRel, Ordering::Acquire)
+            .map_err(|_| SetDefaultError {
+                msg: "default locality was already set!",
+            })?;
+
+        let locality = L::current as *mut ();
+
+        DEFAULT_LOCALITY
+            .compare_exchange(
+                ptr::null_mut(),
+                locality,
+                Ordering::AcqRel,
+                Ordering::Acquire,
+            )
+            .map_err(|_| SetDefaultError {
+                msg: "default locality was already set!",
+            })?;
+        Ok(())
+    }
+
+    pub fn reset() {
+        MAX_LOCALITIES.store(0, Ordering::Acquire);
+        DEFAULT_LOCALITY.store(ptr::null_mut(), Ordering::Acquire);
```

Comment:
i think these should be `Release`? i am also a little iffy about whether this is even a good idea, or if it can be racy (e.g. max localities and the default locality ptr_both_ need to be set in one atomic operation...)

Reply:
this should be hidden behind a `#[cfg(test)]` to be honest, and yeah `Release` should be good enough.